subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -85,61 +88,101 @@ public DefaultPluginRequestApplicator(
-                applyPlugin(result.request, result.found.getPluginId(), new Runnable() {
-                    @Override
-                    public void run() {
-                        result.found.execute(new PluginResolveContext() {
-                            @Override
-                            public void addLegacy(PluginId pluginId, Object dependencyNotation) {
-                                pluginApplyActions.add(target -> applyLegacyPlugin(target, result, pluginId));
-                                scriptHandler.addScriptClassPathDependency(dependencyNotation);
-                            }
-
-                            @Override
-                            public void add(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                            }
-
-                            @Override
-                            public void addFromDifferentLoader(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                                pluginImplsFromOtherLoaders.put(result, plugin);
-                            }
-                        });
-                        String pluginVersion = result.found.getPluginVersion();
-                        if (pluginVersion != null) {
-                            pluginVersionTracker.setPluginVersionAt(
-                                classLoaderScope,
-                                result.found.getPluginId().getId(),
-                                pluginVersion
-                            );
+            applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplsFromOtherLoaders, results);
+        }
+        defineScriptHandlerClassScope(scriptHandler, classLoaderScope, pluginImplsFromOtherLoaders.values());
+    }
+
+    private void resolveAutoAppliedPlugins(PluginRequests autoAppliedPlugins, ScriptHandlerInternal scriptHandler, ClassLoaderScope classLoaderScope, PluginResolver effectivePluginResolver, List<Consumer<PluginManagerInternal>> pluginApplyActions) {
+        ((MutableClassLoaderScope) classLoaderScope).mutate(new MutableClassLoaderScope.MutateActor() {
+            private List<Result> autoAppliedPluginRequests;
+
+            @Override
+            public boolean requiresMutation() {
+                final PluginResolver alreadyOnClasspathIgnoringPluginResolver = new AlreadyOnClasspathIgnoringPluginResolver(effectivePluginResolver,
+                    new ClassloaderBackedPluginDescriptorLocator(scriptHandler.getClassLoader())
+                );
+
+                autoAppliedPluginRequests = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);
+
+                return !autoAppliedPluginRequests.isEmpty();
+            }
+
+            @Override
+            public void mutate(ClassLoaderScope scope) {
```

Comment:
This interface doesn't feel quite right anymore.  For instance, the `MutableClassLoaderScope` always delegates to the child.  We pass in the child classloaderScope here, but then we never use it and we just take advantage of the delegation on the parent.  This interface made sense before all the delegation, because it made it more explicitly something to be used while building the classloaderScope.  With the delegation, we've effectively made it safe to mutate the classloaderScope without unlocking things, so the previous concern about making this only something that could be done while building/locking the scope is not as important now.  I think we could just get rid of the `MutateActor` and have a `mutate(String childSuffix)` method to call when you need to mutate the scope after locking, which creates the child and updates the delegation.  

So this becomes:

```
    private void resolveAutoAppliedPlugins(PluginRequests autoAppliedPlugins, ScriptHandlerInternal scriptHandler, ClassLoaderScope classLoaderScope, PluginResolver effectivePluginResolver, List<Consumer<PluginManagerInternal>> pluginApplyActions) {
        final PluginResolver alreadyOnClasspathIgnoringPluginResolver = new AlreadyOnClasspathIgnoringPluginResolver(effectivePluginResolver,
            new ClassloaderBackedPluginDescriptorLocator(scriptHandler.getClassLoader())
        );
        List<Result> autoAppliedPluginRequests = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);
        if (!autoAppliedPluginRequests.isEmpty()) {
            ((MutableClassLoaderScope) classLoaderScope).mutate("-plugins");
            scriptHandler.dropResolvedClassPath();

            final Map<Result, PluginImplementation<?>> pluginImplementations = newLinkedHashMap();
            applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplementations, autoAppliedPluginRequests);
            defineScriptHandlerClassScope(scriptHandler, classLoaderScope, pluginImplementations.values());
        }
    }
```
Or something like that.  WDYT?

Reply:
I like this approach. I've added it.